benches/dapol.rs

```diff
@@ -1,175 +1,357 @@
+            1
+        } else {
+            max_thread_count >> 2
+        };
+
+        for i in (step..max_thread_count).step_by(step as usize) {
+            tc.push(i);
+        }
+
+        tc.push(max_thread_count);
+
+        tc
+    };
+
+    for h in TREE_HEIGHTS.into_iter() {
+        for t in thread_counts.iter() {
+            for u in NUM_USERS.into_iter() {
+                let mut ndm_smt = Option::<NdmSmt>::None;
+
+                // Many of the statistics tracked by `jemalloc` are cached.
+                // The epoch controls when they are refreshed.
+                // We care about measuring ndm_smt so we refresh before it's construction
+                e.advance().unwrap();
+                let before = alloc.read().unwrap();
+
+                let max_users_for_height = 2_u64.pow((h - 1) as u32);
+
+                if u > max_users_for_height {
+                    break;
+                }
+
+                let tup: (Height, MaxThreadCount, u64) =
+                    (Height::from(h), MaxThreadCount::from(*t), u);
+
+                // tree build compute time
+                group.bench_with_input(
+                    BenchmarkId::new(
+                        "build_tree",
+                        format!("{:?}/{:?}/NUM_USERS: {:?}", tup.0, tup.1, tup.2),
+                    ),
+                    &tup,
+                    |bench, tup| {
+                        bench.iter(|| {
+                            ndm_smt = Some(setup::build_ndm_smt(tup.clone()));
+                        });
+                    },
+                );
+
+                e.advance().unwrap();
+                let after = alloc.read().unwrap();
+
+                // mem used is the difference between the 2 measurements
+                let diff = after - before;
+
+                // tree build file size
+                let tree_build_file_size = setup::serialize_tree(
+                    &ndm_smt.as_ref().expect("Tree not found"),
+                    PathBuf::from("./target"),
+                );
+
+                let tree_build = Metrics {
+                    variable: Variable::TreeBuild,
+                    mem_usage: setup::bytes_as_string(diff),
+                    file_size: tree_build_file_size,
+                };
+
+                println!("\n{:?}\n", tree_build);
+            }
+        }
     }
 
-    group.finish();
+    group.finish()
 }
 
-fn generate_proof(c: &mut Criterion) {
-    let mut group = c.benchmark_group("prove");
+fn bench_generate_proof(c: &mut Criterion) {
+    let e = epoch::mib().unwrap();
+    let alloc = stats::allocated::mib().unwrap();
+
+    let mut group = c.benchmark_group("dapol");
     group.sample_size(10);
+    // `SamplingMode::Flat` is used here as that is what Criterion recommends for long-running benches
+    // https://bheisler.github.io/criterion.rs/book/user_guide/advanced_configuration.html#sampling-mode
+    group.sampling_mode(SamplingMode::Flat);
```

Comment:
These can be changed to better suit the proof generation algorithm. It's not long-running so we don't need the flat sampling mode, and we can do more sample sizes.

Reply:
I thought so as well, but I wanted to hear what you did

Reply:
I increased the sample size to 20, because it gets super slow at 100k users (32 height). Probably because I set `SamplingMode::Auto`, which is defaulting to `SamplingMode::Linear`, which is slow